File: php/src/Advertisements/Advertisement/Domain/Services/AdvertisementSecurityService.php
<?php
declare(strict_types=1);

namespace Demo\App\Advertisements\Advertisement\Domain\Services;

use Demo\App\Advertisements\Advertisement\Domain\Advertisement;
use Demo\App\Advertisements\Shared\ValueObjects\UserId;
use Demo\App\Advertisements\User\Domain\Exceptions\AdminWithIncorrectCivicCenterException;
use Demo\App\Advertisements\User\Domain\Exceptions\UnauthorizedUserException;
use Demo\App\Advertisements\User\Domain\Exceptions\UserNotFoundException;
use Demo\App\Advertisements\User\Domain\UserRepository;

class AdvertisementSecurityService
{
    public function __construct(
        private UserRepository $userRepository,
    ) {}

    /**
     * @throws UserNotFoundException
     * @throws AdminWithIncorrectCivicCenterException
     */
    public function verifyAdminUserCanManageAdvertisement(UserId $securityUserId, Advertisement $advertisement): void
    {
        $adminUser = $this->userRepository->findAdminById($securityUserId);
        if (!$adminUser) {
            throw UserNotFoundException::asAdmin();
        }

        if (!$adminUser->civicCenterId()->equals($advertisement->civicCenterId())) {
            throw AdminWithIncorrectCivicCenterException::differentCivicCenterFromMember();
        }
    }

    /**
     * @throws UnauthorizedUserException
     * @throws UserNotFoundException
     */
    public function verifyMemberUserCanManageAdvertisement(UserId $securityUserId, Advertisement $advertisement): void
    {
        $memberUser = $this->userRepository->findMemberByIdOrFail($securityUserId);
        if (!$memberUser) {
            throw UserNotFoundException::asMember();
        }

        if (!$advertisement->memberId()->equals($securityUserId)) {
            throw UnauthorizedUserException::build();
        }
    }
}

File: kotlin/src/main/kotlin/advertisements/advertisement/domain/services/AdvertisementSecurityService.kt
package advertisements.advertisement.domain.services

import advertisements.advertisement.domain.Advertisement
import advertisements.shared.value_object.UserId
import advertisements.user.domain.UserRepository
import advertisements.user.domain.exceptions.AdminWithIncorrectCivicCenterException
import advertisements.user.domain.exceptions.UnauthorizedUserException
import advertisements.user.domain.exceptions.UserNotFoundException

class AdvertisementSecurityService(private val userRepository: UserRepository) {

    fun verifyAdminUserCanManageAdvertisement(securityUserId: UserId, advertisement: Advertisement) {
        val adminUser = userRepository.findAdminById(securityUserId)
            ?: throw UserNotFoundException.asAdmin()

        if (!adminUser.civicCenterId().equals(advertisement.civicCenterId)) {
            throw AdminWithIncorrectCivicCenterException.differentCivicCenterFromMember()
        }
    }

    fun verifyMemberUserCanManageAdvertisement(securityUserId: UserId, advertisement: Advertisement) {
        val memberUser = userRepository.findMemberById(securityUserId)
            ?: throw UserNotFoundException.asMember()

        if (!advertisement.memberId.equals(securityUserId)) {
            throw UnauthorizedUserException.build()
        }
    }
}
